### PR TITLE
fix(leader-election): keepalive/campaign/lease/watcher recovery and validation (FIB-173, FIB-175, FIB-176, FIB-177, FIB-178, FIB-182)

### PR DESCRIPTION
**Pull request by kyonRay**
## CertiK preliminary findings FIB-173/175/176/177/178/182 (bcos-leader-election)

Six Minor findings from the CertiK preliminary review (2026-06-11), one commit each. All are defensive/robustness fixes in the etcd-backed leader-election path.

### FIB-176 — invalid `leaseTTL` not validated before deriving timers
`leaseTTL` (unsigned) was stored unchecked: `0` builds a 0 ms campaign timer and `leaseTTL()-1` underflows to a huge value passed to `etcd::KeepAlive`; `1` yields a 0 s keepalive interval. Fix: add `c_minLeaseTTL=2` + `clampLeaseTTL()` and clamp in the ctor (WARNING when clamped).

### FIB-182 — watcher reconnect snapshot didn't prune deleted keys
`WatcherConfig::fetchLeadersInfo()` only added/updated keys from the `ls()` snapshot and never removed keys absent from it, so a leader key deleted while the watcher was disconnected stayed cached indefinitely (still visible via `getLeaderByKey()`/`getAllLeaders()`; delete handlers never fired). Fix: treat the snapshot as authoritative — under the state lock, erase cached keys absent from the snapshot and fire `onMemberDeleted()` for each outside the lock.

### FIB-173 — stale self-leader state after keepalive failure
`onKeepAliveException()` restarted the campaign timer but did not clear the cached self-leader, so the next campaign short-circuited on the stale `getLeader()==self` and the node kept master state until an external watcher event. Fix: clear the cached self-leader (`resetLeader(nullptr)`, made public) and notify the upper layer to leave master mode before restarting. The dead keepalive is released via `m_keepAlive.reset()` rather than `Cancel()`, because `Cancel()` joins the refresh thread that invokes this very handler (self-deadlock).

### FIB-175 — unbounded blocking etcd lease grant
`grantLease()` called `leasegrant(...).get()` with no timeout/cancellation; a stalled/unreachable etcd wedged the timer-driven campaign thread and `stop()` could not interrupt it. Fix: arm a client-level grpc timeout (`set_grpc_timeout(m_etcdOpTimeout)`, ~one lease period) once in the constructor. In etcd-cpp-apiv3 v0.15.4 this deadline is applied on the completion-queue wait (`cq_.AsyncNext(grpc_deadline())`) of unary RPCs (leasegrant/txn/get) — not a stream `context.set_deadline` — so the blocking `get()` returns a not-ok response when the deadline elapses (callers already fall back to backup + restart the timer). The watch stream pumps an unbounded `cq_.Next()` loop and is unaffected; only a keepalive refresh becomes bounded by one lease period, which is benign (a refresh that cannot complete within a full TTL means the lease is already lost). The duration is built as `chrono::seconds(leaseTTL)` so a large TTL cannot overflow the unsigned multiply before the cast. The same client-level bound also covers the previously-unbounded `txn`/`get` blocking calls.

### FIB-177 — malformed etcd leader value had no bounded recovery
`checkAndUpdateLeaderKey()` decoded the leader value via `createMember()`; a malformed value threw, the startup path only logged it, and nothing cleared/quarantined the bad key, so every fetch re-failed with no progress. Fix: wrap the decode in try/catch; on failure clear the leader and trigger a bounded re-campaign. A `catch (...)` fallback applies the same recovery to non-std throwables so an exotic exception cannot escape into the watcher callback.

### FIB-178 — `campaignLeader()` exception path didn't reschedule recovery
The catch-all logged and cancelled keepalive but returned without restarting the timer or switching to backup, so a transient exception left the node dependent on an external trigger to campaign again. Fix: mirror the non-exception failure paths — `tryToSwitchToBackup()` + restart the campaign timer.

Refs: `audit/findings/FIB-173.md`, `FIB-175.md`, `FIB-176.md`, `FIB-177.md`, `FIB-178.md`, `FIB-182.md`.

Verification: changes pass `clang-format`; type signatures checked against the etcd/cpprestsdk headers, and the FIB-175 `grpc_timeout` semantics confirmed against etcd-cpp-apiv3 v0.15.4 source (the deadline binds the unary completion-queue wait, not the watch stream). Unit tests are not included for this module (it is constructor-bound to a live `etcd::Client` with no in-tree mock); behavior is gated by the CI build (which provides etcd/gRPC) plus review. Each fix is surgical and self-contained.

### CODE REVIEW

File: bcos-leader-election/src/LeaderElection.cpp [hunk 1]
```diff
@@ -71,7 +73,48 @@ void LeaderElection::stop()
...
+            try
+            {
+                promise->set_exception(std::current_exception());
+            }
+            catch (...)
+            {
+                // promise already satisfied / broken; nothing more to do.
+            }
+        }
+    }).detach();
```
Comment (ywy2090) [inline bcos-leader-election/src/LeaderElection.cpp]: ⚠️ [IMPORTANT] 超时后主线程返回 `{false,0}`，但 detached worker 仍在 `leasegrant().get()` 阻塞；campaign 定时器每次重试都会再 spawn 一个线程。etcd 长时间故障时可能线程堆积。建议：串行化 grant（mutex/atomic 单飞）、用有界线程池、或在超时后让 worker 可被 `stop()` 取消/丢弃。

**Review by ywy2090 — CHANGES_REQUESTED** [review]
## Overview

PR #5265 针对 bcos-leader-election 六项 Minor 审计项：leaseTTL 校验、watcher 重连快照修剪、keepalive 失败后清除自引用 leader、etcd leasegrant 超时、畸形 leader value 恢复、campaign 异常路径自愈。整体方向正确，FIB-173/176/177/178/182 改动合理。

## Concerns

- **FIB-175 detached thread 堆积**：`grantLease()` 超时后主线程返回，但 detached worker 仍在 `leasegrant().get()` 上阻塞；campaign 定时器周期性重试会不断 spawn 新线程。etcd 长时间不可达时可能耗尽线程资源，属于对原「阻塞 campaign 线程」问题的部分缓解但未完全消除。
- FIB-175 未在 `stop()` 时取消进行中的 leasegrant；节点关闭期间 detached worker 可能短暂存活（通常可接受）。
- 模块无单元测试（PR 已说明 etcd 依赖），依赖 CI 构建与人工审查。
- `.gitignore` 附带 `.worktrees/` 与修复无关。

## Verdict

Found: 0 critical, 1 important, 0 suggestions

**Request changes**

**Review by morebtcg — COMMENTED** [review]
## Review of PR #5265: fix(leader-election): keepalive/campaign/lease/watcher recovery and validation (FIB-173,175,176,177,178,182)

### 总体评价
⚠️ **Approve with suggestions** — 六个修复整体方向正确，但 FIB-175 的 detached thread 方案有一个潜在的资源泄漏窗口需要关注。

---

### FIB-176: leaseTTL 校验

#### `bcos-leader-election/src/CampaignConfig.h` & `.cpp`
**✅ `c_minLeaseTTL=2` + `clampLeaseTTL()`** — 防御性设计合理。0 和 1 的 leaseTTL 确实会导致退化行为（0ms 定时器 / 0s keepalive 间隔）。

**小建议:** 警告日志在构造函数中打印，但 `m_leaseTTL` 赋值在日志之后。如果 `_leaseTTL < c_minLeaseTTL`，日志打印的是原始 `_leaseTTL`（正确），但之后 `m_leaseTTL = clampLeaseTTL(_leaseTTL)` 时 `_leaseTTL` 已被修改。当前代码没问题是因为日志在 clamp 之前。但为了代码清晰，建议将日志和 clamp 放在一起或将原始值存为局部变量。

---

### FIB-177: 畸形 etcd leader value 恢复

#### `bcos-leader-election/src/CampaignConfig.cpp`
**✅ try/catch 包裹 `createMember()`** — 正确。之前 decode 失败只打日志，不清理异常 key，导致每次 fetch 都重复失败。

**建议:** `catch (std::exception const& e)` 这里只捕获了标准异常。如果 `createMember` 抛出非标准异常（例如某种协程取消异常），这里不会捕获。建议加一个 `catch (...)` 作为兜底，或确认 `createMember` 的异常规范。

---

### FIB-182: watcher 重连快照未清理已删除 key

#### `bcos-leader-election/src/WatcherConfig.cpp`
**✅ 快照权威化处理** — 正确。`fetchLeadersInfo()` 是 etcd watcher 重连后的全量同步，应该以快照为准删除本地缓存的过期 key。

**注意:** `prunedMembers` 在锁外触发 `onMemberDeleted` 回调，防止了在回调中获取同一把锁导致的死锁。这是好的。

**小建议:** 如果 `values` 很大（大量 leader key），构造 `std::set<std::string> snapshotKeys` 会有内存开销。对于 leader election 场景来说 key 数量通常不多（几十个），所以影响不大。但如果未来扩展到其他用途，可以考虑直接在 `m_keyToLeader` 上做双指针遍历（`m_keyToLeader` 按 key 排序的话）。

---

### FIB-175: 有界 etcd lease grant

#### `bcos-leader-election/src/LeaderElection.cpp`
**⚠️ 方案正确但有一个资源泄漏窗口：** detached thread + promise/future 方案避免了 `std::async` 的析构阻塞问题，设计思路是合理的。

**潜在问题:** detached thread 可能在 `LeaderElection` 对象被销毁后继续运行。线程捕获了 `m_etcdClient`（shared_ptr），所以 etcd client 不会被提前释放——这是安全的。但如果 `m_etcdClient` 本身在析构时关闭了底层连接，而 detach 的线程仍在 `leasegrant(...).get()` 中阻塞，可能导致：
1. 线程访问已关闭的连接（取决于 etcd client 实现是否线程安全地处理关闭）
2. 线程在进程退出时被强制终止，留下未完成的 promise

**建议:** 在 `LeaderElection::stop()` 中加入对 in-flight lease grant 的追踪，确保 stop 后不再有 detach 线程访问 client。可以考虑用一个 `std::atomic<bool> m_stopping` 标志，让 worker 线程在 `leasegrant` 返回后检查。或者使用 `etcd::Client` 的取消机制（如果提供的话）。

**另一个方案（供参考）:** 如果 `etcd::Client` 提供了带超时的异步 API（如 `leasegrant(ttl).wait_for(timeout)`），就不需要 detached thread。但 PR 中提到 `pplx::task` 没有 wait-with-timeout，所以当前方案是合理的权衡。

**小问题:** promise 的 `set_exception` 失败时被静默吞掉（nested try/catch 注释说 "promise already satisfied"）。虽然这在 `wait_for` 超时后是预期行为，但如果 promise 因为其他原因失败，问题会被隐藏。建议至少加一条 DEBUG 日志。

---

### FIB-173: keepalive 失败后残留 self-leader 状态

#### `bcos-leader-election/src/LeaderElection.cpp`
**✅ `m_keepAlive.reset()` 而非 `Cancel()`** — PR 描述中已说明：`Cancel()` 会 join 当前线程（keepAlive 的 refresh 线程），导致自死锁。`reset()` 释放引用是正确的。注释已充分说明原因。

**✅ `m_config->resetLeader(nullptr)` + `m_onCampaignHandler(false, ...)`** — 正确。必须先清理本地 leader 状态再通知上层，否则下一轮 campaign 会短路。

**注意:** `RecursiveGuard l(m_mutex)` 是递归锁——请确认此处确实需要递归语义，还是可以用普通 `std::lock_guard`。如果递归锁只是因为历史原因，改为非递归锁可以减少死锁风险。

---

### FIB-178: campaignLeader 异常路径不触发恢复

#### `bcos-leader-election/src/LeaderElection.cpp`
**✅ catch-all 中增加 `tryToSwitchToBackup()` + timer restart** — 与非异常失败路径保持一致。

---

### 总体建议

1. **单元测试:** PR 描述提到 "Unit tests are not included for this module (constructor-bound to live etcd::Client with no in-tree mock)"。虽然理解测试困难，但至少 FIB-176（leaseTTL clamp）和 FIB-177（malformed value decode）是纯逻辑，可以独立测试。建议后续为这些不依赖 etcd 的修复补充单元测试。

2. **FIB-175 detached thread:** 这是整个 PR 中风险最高的改动。建议在合并前确保 stop() 路径不会导致 detached thread 访问已销毁的 etcd client。

**Comment by kyonRay** [comment]
## FIB-175 已按 review 重做（force-push，FIB-175 = `8fca753ab`）

把 detached-thread + `std::promise` 方案整套删掉，改为构造时设一次 client 级 `set_grpc_timeout(m_etcdOpTimeout)`（约一个 lease 周期）。

核对 etcd-cpp-apiv3 **v0.15.4** 源码确认语义安全：该超时作用在一元 RPC 的**完成队列等待** `cq_.AsyncNext(grpc_deadline())` 上（leasegrant/txn/get），并非 stream 的 `context.set_deadline`；**watch** 主循环走无界 `cq_.Next()`，不受影响；仅 **KeepAlive refresh** 变有界（一个 lease 周期，良性——一次 refresh 连一个完整 TTL 都刷不回来，说明租约本就没了）。

### @ywy2090
你点出的 detached worker 堆积、超时后不可取消、`stop()` 期残留，现在**全部消失**——因为已经没有线程了。stalled etcd 会让 `leasegrant().get()` 在约一个 lease 周期后自行返回 not-ok，调用方既有的 backup + 重启定时器逻辑接管。顺带把 `txn`/`get` 的同类无界 `.get()` 也一并封了边界。请重新 review。

### @morebtcg
- **FIB-175**：采纳你「若有带超时的 API 就不必开线程」的方向，thread/promise 全删，资源泄漏窗口随之消失。
- **FIB-177**：已加 `catch(...)` 兜底（`81eac6d46`），非标准异常走同样的 drop-leader + recampaign 恢复。
- **FIB-176 日志顺序 / FIB-182 `std::set`**：如你所言当前实现无问题，按选举场景规模保留现状。
- **FIB-173 递归锁**：属 FIB-169 既有设计，本 PR 不改锁语义。

PR 描述的 FIB-175 段落也已同步更新。

**Review by morebtcg — COMMENTED** [review]
## Review 确认 ✅

已逐项复核 FIB-173/175/176/177/178/182 的修复，确认之前提出的问题均已解决：

### FIB-176 — leaseTTL 校验
- ✅ `c_minLeaseTTL=2` + `clampLeaseTTL()` 防止 0 产生 0ms campaign timer 以及 `leaseTTL()-1` 下溢
- ✅ 构造函数中 WARNING 日志记录 clamp 行为

### FIB-182 — watcher 重连快照清理
- ✅ 将 `ls()` 快照视为权威数据源，在 state lock 内擦除快照中不存在的 key
- ✅ `onMemberDeleted()` 回调在 lock 外触发，避免自死锁
- ✅ 日志增加 `prunedSize` 便于排查

### FIB-173 — keepalive 失败后残留 master 状态
- ✅ `onKeepAliveException()` 中先 `reset()` 释放 keepAlive（避免 `Cancel()` 自死锁），再 `resetLeader(nullptr)` 清除残留 leader 指针，然后通知上层退出 master 模式，最后重启 campaign
- ✅ `resetLeader` 已从 protected 提升为 public

### FIB-175 — etcd lease grant 无超时
- ✅ `std::chrono::seconds(leaseTTL)` → `std::chrono::milliseconds` 构造 `m_etcdOpTimeout`，无溢出风险
- ✅ 构造函数中通过 `set_grpc_timeout` 武装 etcd client，使阻塞 `get()` 可超时返回

### FIB-177 — 畸形 leader value
- ✅ `createMember()` 包裹 try-catch（std::exception + catch-all），与 version==0 恢复路径一致
- ✅ drop stale leader + 触发重新 campaign

### FIB-178 — campaignLeader 异常路径未恢复
- ✅ catch 块中补充 `tryToSwitchToBackup()` + `m_campaignTimer->restart()`，不再让节点卡死

未发现遗留问题，LGTM。

**Review by morebtcg — APPROVED** [review]
